lib/dynflow/tasks/export_task.rb

```diff
@@ -0,0 +1,146 @@
+require 'rake'
+require 'rake/tasklib'
+
+module Dynflow
+  module Tasks
+    class Export < ::Rake::TaskLib
+
+      attr_accessor :name, :world
+      attr_writer :task_days
+      attr_writer :task_file
+      attr_writer :task_format
+      attr_writer :task_search
+      attr_writer :plans
+
+      def initialize(name = :export_tasks)
+        @name = name
+        @world = nil
+        @plans = nil
+
+        @task_file = nil # Filename of the resulting file
+        # Search options
+        @task_days    = nil # Number of days
+        @task_format  = nil # One of html, csv, json
+        @task_ids     = nil # Comma separated list of task ids
+        @task_states  = nil # Comma separated list of task states
+        @task_results = nil # Comma separated list of task results
+        yield self if block_given?
+        define
+      end
+
+      def define
+        desc 'Export tasks'
+        task @name do
+          export
+        end
+        self
+      end
+
+      def task_file
+        ENV['TASK_FILE'] ||
+          @task_file ||
+          "/tmp/task-export-#{Time.now.to_i}.#{task_format == 'csv' ? 'csv' : 'tar.gz'}"
+      end
+
+      def task_format
+        ENV['TASK_FORMAT'] || @task_format || 'html'
+      end
+
+      def task_days
+        ENV['TASK_DAYS'] || @task_days
+      end
+
+      def task_states
+        return @task_states if @task_states
+        @task_states = ENV['TASK_STATES'].split(',') if ENV['TASK_STATES']
+      end
+
+      def task_results
+        return @task_results if @task_results
+        @task_results = ENV['TASK_RESULTS'].split(',') if ENV['TASK_RESULTS']
+      end
+
+      def task_ids
+        return @task_ids if @task_ids
+        @task_ids = ENV['TASK_IDS'].split(',') if ENV['TASK_IDS']
+      end
+
+      private
+
+      def export
+        if plans.empty?
+          puts("Nothing to export, exiting")
+          return
+        end
+
+        puts "Exporting #{plans.count} tasks"
+        content = case task_format
+                  when 'html'
+                    ::Dynflow::Exporters::Tar.full_html_export plans
+                  when 'json'
+                    ::Dynflow::Exporters::Tar.full_json_export plans
+                  when 'csv'
+                    ::Dynflow::Exporters::CSV.new
+                      .add_many(plans).result
+                  else
+                    raise "Unknown export format '#{format}'"
+                  end
+        File.write(task_file, content)
+        puts "Exported tasks as #{task_file}"
```

Comment:
It indeed puts everything into memory and then dumps it. The main concern of this PR was to move the Dynflow specific bits and pieces from Foreman Tasks, so the same logic could be used for exporting tasks from Foreman Tasks as well as Smart Proxy, while being easily extensible.

But agreed, performance here isn't exactly great, so I'll see what can be done about it.